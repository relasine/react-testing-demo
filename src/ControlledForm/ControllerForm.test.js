import React from "react";
import { shallow } from "enzyme";

import ControlledForm from "./ControlledForm";

import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("ControlledForm", () => {
  let wrapper;
  let mockSubmit;

  beforeEach(() => {
    mockSubmit = jest.fn();
    wrapper = shallow(<ControlledForm submit={mockSubmit} />);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("handleChange", () => {
    it("should call setState on title", () => {
      const mockEvent = {
        target: {
          name: "title",
          value: "test"
        }
      };

      const expected = {
        title: "test",
        description: "",
        submitActive: false
      };
      wrapper.instance().handleChange(mockEvent);

      expect(wrapper.state()).toEqual(expected);
    });

    it("should call setState on description", () => {
      const mockEvent = {
        target: {
          name: "description",
          value: "test"
        }
      };

      const expected = {
        title: "",
        description: "test",
        submitActive: false
      };
      wrapper.instance().handleChange(mockEvent);

      expect(wrapper.state()).toEqual(expected);
    });

    it("should call checkFields", () => {
      const spy = jest.spyOn(wrapper.instance(), "checkFields");

      wrapper.instance().forceUpdate();

      const mockEvent = {
        target: {
          name: "description",
          value: "test"
        }
      };

      const expected = true;

      wrapper.instance().handleChange(mockEvent);

      expect(spy).toHaveBeenCalled();
    });
  });

  it("should call handleChange on title change with the correct params", () => {
    const spy = jest.spyOn(wrapper.instance(), "handleChange");

    wrapper.instance().forceUpdate();

    const mockEvent = {
      target: {
        name: "title",
        value: "test"
      }
    };

    wrapper.find(".title-input").simulate("change", mockEvent);

    expect(spy).toHaveBeenCalledWith(mockEvent);
  });

  it("should call handleChange on description change with the correct params", () => {
    const spy = jest.spyOn(wrapper.instance(), "handleChange");

    wrapper.instance().forceUpdate();

    const mockEvent = {
      target: {
        name: "description",
        value: "test"
      }
    };

    wrapper.find(".description-input").simulate("change", mockEvent);

    expect(spy).toHaveBeenCalledWith(mockEvent);
  });

  describe("handleSubmit", () => {
    it("should call preventDefault", () => {
      const mockPreventDefault = jest.fn();

      const mockEvent = {
        preventDefault: mockPreventDefault
      };

      wrapper.instance().handleSubmit(mockEvent);

      expect(mockPreventDefault).toHaveBeenCalled();
    });

    it("should return if submitActive is false", () => {
      const mockPreventDefault = jest.fn();

      const mockEvent = {
        preventDefault: mockPreventDefault
      };

      const spy = jest.spyOn(wrapper.instance(), "handleSubmit");
      wrapper.instance().forceUpdate();

      wrapper.instance().handleSubmit(mockEvent);

      expect(spy).toReturn();
    });

    it("should call submit with the correct params", () => {
      wrapper.setState({
        title: "test title",
        description: "test description",
        submitActive: true
      });

      const expected = {
        title: "test title",
        description: "test description"
      };

      const mockPreventDefault = jest.fn();

      const mockEvent = {
        preventDefault: mockPreventDefault
      };

      wrapper.instance().handleSubmit(mockEvent);

      expect(mockSubmit).toHaveBeenCalledWith(expected);
    });

    it("should call handleSubmit on submit with the correct params", () => {
      const spy = jest.spyOn(wrapper.instance(), "handleSubmit");
      wrapper.instance().forceUpdate();

      const mockPreventDefault = jest.fn();

      const mockEvent = {
        preventDefault: mockPreventDefault
      };

      wrapper.find(".controlled-form").simulate("submit", mockEvent);

      expect(spy).toHaveBeenCalledWith(mockEvent);
    });
  });

  describe("checkFields", () => {
    it("should setState if it meets conditions", () => {
      wrapper.setState({
        title: "test title",
        description: "test description",
        submitActive: false
      });

      wrapper.instance().checkFields();

      expect(wrapper.state().submitActive).toEqual(true);
    });

    it("should setState if it does not meet conditions", () => {
      wrapper.setState({
        title: "test title",
        description: "",
        submitActive: true
      });

      wrapper.instance().checkFields();

      expect(wrapper.state().submitActive).toEqual(false);
    });
  });
});
