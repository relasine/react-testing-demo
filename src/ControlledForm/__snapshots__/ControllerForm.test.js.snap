// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledForm should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ControlledForm
    submit={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="text-input title-input"
          name="title"
          onChange={[Function]}
          type="text"
          value=""
        />,
        <input
          className="text-input description-input"
          name="description"
          onChange={[Function]}
          type="text"
          value=""
        />,
        <button
          className="submit-btn false"
        >
          submit
        </button>,
      ],
      "className": "controlled-form",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "text-input title-input",
          "name": "title",
          "onChange": [Function],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "text-input description-input",
          "name": "description",
          "onChange": [Function],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "submit",
          "className": "submit-btn false",
        },
        "ref": null,
        "rendered": "submit",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="text-input title-input"
            name="title"
            onChange={[Function]}
            type="text"
            value=""
          />,
          <input
            className="text-input description-input"
            name="description"
            onChange={[Function]}
            type="text"
            value=""
          />,
          <button
            className="submit-btn false"
          >
            submit
          </button>,
        ],
        "className": "controlled-form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "text-input title-input",
            "name": "title",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "text-input description-input",
            "name": "description",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "submit",
            "className": "submit-btn false",
          },
          "ref": null,
          "rendered": "submit",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
